### PR TITLE
Handle multi-master flannel migration

### PR DESCRIPTION
**Pull request by caseydavenport**
## Description

The controller doesn't properly handle multi-control-plane nodes.

Should fix https://github.com/projectcalico/kube-controllers/issues/495

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
Fix flannel migration for clusters with multiple control-plane nodes. 
```

### CODE REVIEW

File: pkg/controllers/flannelmigration/migration_controller.go
Comment (song-jiang) [inline pkg/controllers/flannelmigration/migration_controller.go:386]: Not sure if we need this change since there should always be a single node running migration controller.

File: pkg/controllers/flannelmigration/migration_controller.go
Comment (song-jiang) [reply pkg/controllers/flannelmigration/migration_controller.go:442]: @caseydavenport The idea is if we start migrating master nodes and hit any unexpected issue, there is a risk that nodes lost network connections with kube-apiserver and no further migration is possible after that. To have a deterministic order of nodes list, once migration process reaches master node, we have confidence that majority of nodes has been migrated successfully.

**Review by song-jiang — COMMENTED** [review]
Thanks @caseydavenport for a quick fix!

**Comment by seh** [comment]
It looks like the CI build failed _go vet_ due to extraneous use of `Infof` with no format directives.

Will someone be able to address the comments and get this merged soon?

File: pkg/controllers/flannelmigration/migration_controller.go
Comment (seh) [inline pkg/controllers/flannelmigration/migration_controller.go:437]: _go vet_ complained that this format string is missing a directive. Perhaps:
```suggestion
		log.Infof("Adding control plane nodes to end of queue: %v.", masterNodes)
```

File: pkg/controllers/flannelmigration/migration_controller.go
Comment (caseydavenport) [reply pkg/controllers/flannelmigration/migration_controller.go:386]: Yeah, true. This is only ever going to be one item long.

Comment (caseydavenport) [reply pkg/controllers/flannelmigration/migration_controller.go:386]: I will leave it like this though, just in case something weird happens we'll handle it more gracefully.

**Comment by caseydavenport** [comment]
@seh sorry, I was on PTO last week. PR is updated now and I think good to go.

**Comment by caseydavenport** [comment]
Waiting on @song-jiang for a final approval

**Comment by seh** [comment]
There's no need to apologize, Casey. I've been working around the problem by waiting until the migration controller finishes, then deleting an EC2 instances hosting a master node that's still cordoned. It causes a delay for a few minutes, but doesn't cause any downtime for the Kubernetes API.

The latest commit looks good to me. I'd love to have a released container image including this fix available for when I run my upgrades of eleven clusters (planned so far).

**Review by song-jiang — APPROVED** [review]
LGTM.

**Comment by caseydavenport** [comment]
If you'd like to try this out prior to release, I believe it should be available in this dev image: `docker push calico/flannel-migration-controller:v3.17.0-0.dev-17-g95f4094-ppc64le`

**Comment by seh** [comment]
I've only run it once so far, but it ran to completion successfully, visiting every node without leaving any of the multiple master nodes cordoned. That looks much better so far. Thank you again.

**Comment by seh** [comment]
@caseydavenport, is this patch included in a release-grade container image? I've been using the one you mentioned above, but I'd prefer to use one less likely to raise eyebrows in code review.

[The migration manifests](https://docs.projectcalico.org/manifests/flannel-migration/migration-job.yaml) are using an image with tag 3.16.1 at present.

**Comment by willemveerman** [comment]
I'm also interested in when this fix might be released - I see it's included in the 3.17.0 rc

Is there a rough ETA for 3.17.0?

**Comment by song-jiang** [comment]
According to tag v3.16.1, this PR fix is included.
```
v3.16.1 …
- 0dfe2a6 Handle multi-master flannel migration
- 78710a3 Fix `kubectl exec` format for migration controller
- 8d2d272 Update migration controller for rancher
- 879727e Update pins
```

**Comment by seh** [comment]
Ah, I was looking at [_v3.16.1 ... master_](https://github.com/projectcalico/kube-controllers/compare/v3.16.1...master). There we can see commit 12a0a98f782dced12d5ab89259b1255d1b130bce. I thought that meant that the "master" branch is ahead of tag _v3.16.1_, and that that commit is not present in _v3.16.1_. Looking today at [the commit history from tag _v3.16.1_](https://github.com/projectcalico/kube-controllers/commits/v3.16.1), now I see that the patch of interest is indeed there.

I apologize for having read this wrong (https://github.com/projectcalico/kube-controllers/pull/496#issuecomment-692704145).

**Comment by seh** [comment]
Note that this patch was not backported to the 3.15 minor version. (It's not present [in version 3.15.3](https://github.com/projectcalico/kube-controllers/commits/v3.15.3).)

**Comment by caseydavenport** [comment]
Sounds like you guys figured it out, but here's the cherry-pick PR which is in the v3.16.1 milestone as further evidence: https://github.com/projectcalico/kube-controllers/pull/503

**Pull request merged**
